wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/wsnext/SidecarBasedInternalEnvironmentProvider.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.api.workspace.server.wsnext;
+
+import java.util.Collection;
+import java.util.Map;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ValidationException;
+import org.eclipse.che.api.core.model.workspace.config.Environment;
+import org.eclipse.che.api.workspace.server.InternalEnvironmentProvider;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.environment.InternalEnvironment;
+import org.eclipse.che.api.workspace.server.spi.environment.InternalEnvironmentFactory;
+import org.eclipse.che.api.workspace.server.wsnext.model.ChePlugin;
+
+/**
+ * Creates {@link InternalEnvironment} from provided environment configuration, then converts it
+ * using {@link InternalEnvironmentConverter} and applies {@link WorkspaceNextApplier} to the
+ * resulting {@code InternalEnvironment} to add sidecar tooling
+ *
+ * @author Oleksandr Garagatyi
+ */
+public class SidecarBasedInternalEnvironmentProvider extends InternalEnvironmentProvider {
+  private final Map<String, WorkspaceNextApplier> workspaceNextAppliers;
+  private final WorkspaceNextObjectsRetriever workspaceNextObjectsRetriever;
+  private final InternalEnvironmentConverter environmentConverter;
+
+  public SidecarBasedInternalEnvironmentProvider(
+      Map<String, WorkspaceNextApplier> workspaceNextAppliers,
+      WorkspaceNextObjectsRetriever workspaceNextObjectsRetriever,
```

Comment:
WorkspaceNextObjectsRetriever - can we avoid usage of workspace next word in class names?

Reply:
This is a class that we already have in master. And as I described [here](https://github.com/eclipse/che/pull/10474#discussion_r203679226) naming is not important at this stage of this PR.
But I agree with you and it is what Gorkem suggested in one of the previous PRs, so you are not along who thinks so.